### PR TITLE
Unblock special commit message

### LINKED ISSUE #6993: Commit Message Text field stays deactivated after fixup commit
## Current behaviour

* Start making a fixup commit
  * Commit message text field is deactivated
* Do the commit
  * Commit window stays open (as configured)
  * Commit message text field stays deactivated
  * Also other controls like checkbox for amend commit stays deactivated
* Close and reopen commit window
  * Commit message text field is activated again

## Expected behaviour

After making a fixup commit all controls, in special the commit message text field, should be activated again

## Workaround

Reopen commit window

## Environment

- GitExtensions version: 3.1.1.6049 (Build 2f872105b4eb42421d9fdafd68e90a9e27d20d96)
- GIT version: git version 2.22.0.windows.1
- OS version: Windows 10 1607

### LINKED ISSUE #7104: Don't block commit message textbox for fixup! squash! commits
## Don't block commit message textbox for fixup! or squash! commits

I feel that changes
> disable changing the commit message and amending a commit unless CommitKind.Normal

introduced in https://github.com/gitextensions/gitextensions/pull/6530 limit the user in some situations which are valid:

- Open commit dialog in fixup mode and change your mind - previously we could've just deleted the text
- Turn current latest commit into a fixup commit.. not possible without typing fixup! manually as amend is disabled
- Open commit dialog in fixup mode, commit part of the files and the make another regular commit (covered by #6993)

I find myself hitting the "readonly wall" quite often in at least two of the above cases.

The proposal would be to revert this behavior to enable more flexible usage scenarios.

## Environment

- Git Extensions 3.2.0
- Build 3010403a6229cafa8a364f60ec2a2ee4517fcb26 (Dirty)
- Git 2.23.0.windows.1
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.3815.0
- DPI 192dpi (200% scaling)

### PR DESCRIPTION
**Pull request by mstv**
Fixes #7104
Fixes #6993

## Proposed changes

- add button to enable editing the special commit message of `fixup!` and `squash!` commits
- add/adapt hotkeys for buttons
- allow `Amend commit'` for any `CommitKind`
- reset `CommitKind` after commiting without closing `FormCommit`
- slightly off topic: add more test cases for empty lines in commit messages (addendum to #6877)

### Before

grafik

### After

grafik
(different message font due to different font settings vs. portable build)
grafik

### (Intermediate with ToolStripItemButton)
grafik

### Before

grafik

### After

grafik

## Test methodology 

- manual tests

## Test environment(s) 

- Git Extensions 3.3.0
- Build 79dee30803b7eb1ae0ff5999ff3bb962a9cacff7
- Git 2.23.0.windows.1
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.3815.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by mstv** [comment]
Added hotkeys also for the `Reset` buttons, @mrexodia.

**Comment by mrexodia** [comment]
Cool, thanks!

**Comment by RussKie** [comment]
> * add/adapt hotkeys for buttons

I propose this is submitted as its own PR, as it will get merged quickly.

> * slightly off topic: add more test cases for empty lines in commit messages (addendum to #6877)

So as this.

> * allow `Amend commit'` for any `CommitKind`

Could you please explain why this is a good change?

> * add button to enable editing the special commit message of `fixup!` and `squash!` commits
> grafik

I do not disagree with the change, but I am not fond of the proposed implementation.
There is no immediate connection (visual or otherwise) between the textbox and the button.

I think we provide a better user experience by rendering a link label on top of the textbox (z-order wise), e.g. something like the following:
image
The text on the link is up for a discussion (it could be "cancel fixup"?).

**Comment by mstv** [comment]
> allow Amend commit' for any CommitKind

https://github.com/gitextensions/gitextensions/issues/7104#issuecomment-528749535: "It's a totally valid scenario to turn the latest commit into a fixup! commit."

> I think we provide a better user experience by rendering a link label on top of the textbox (z-order wise)

Understood. Can it be activated using a shortcut key?

> The text on the link is up for a discussion (it could be "cancel fixup"?).

I'd prefer "Modify this special commit message", perhaps with an addition like "(The warrenty ends here.)". (I know we don't give warrenties and should never do.)
It is *not* cancelling the fixup: The message will not be changed by clicking. Just `CommitKind` is switched to `Normal` in order to store the manually entered commit message on (temporary) close of FormCommit.

**Comment by RussKie** [comment]
> Understood. Can it be activated using a shortcut key?

I would think it can.
https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.label.usemnemonic?view=netframework-4.8#System_Windows_Forms_Label_UseMnemonic

> I'd prefer "Modify this special commit message", perhaps with an addition like "(The warrenty ends here.)".

We can have a tooltip that warns users (e.g. "do it at your own risk" kind of thing)

**Comment by mstv** [comment]
Fiddled around with `[Link]Label` without success. As I explained above, we need a button-like control, not a label which controls only the focus - the documentation (which I read too late) states:
"Pressing ALT + the mnemonic character sets the focus to the control that follows the Label in the tab order."
BTW: The VS 2019 Designer does not understand `FormCommit` completely. And it did not allow to place `[Link]Label` in the area of the message editor.
grafik

**Comment by RussKie** [comment]
I'll try to find some time and fiddle with it

**Comment by mstv** [comment]
I turned the `ToolStripItemButton` into a `Label`, into a `LinkLabel` and finally into a `Button` in dedicated commits.
I think this is a good compromise:
grafik

TODO:
* make ToolTip translatable:
The `TranslationApp` seems to ignore ToolTips. Must I really create a `TranslationString` by hand?
* squash

**Comment by gerhardol** [comment]
> Does it look OK with a dark theme?

It is simple to test with high contrast dark theme (nothing looks OK, But the button should be visible. 
To install a real dark system the only simple way is to use WindowsBlinds. Paid app, but have a trial.
https://github.com/gitextensions/gitextensions/pull/6651#issuecomment-495798173

**Comment by mdonatas** [comment]
Food for thought. From UX perspective it seems weird to make a text-area read only and then place a button to make it editable... for what reason exactly? There is no explanation and adding an explanation would only dig this hole deeper.
Git users know that there is nothing special with `fixup!` and `squash!` they are just guidelines for interactive rebase to automate reordering rows.
If someone is going to do interactive rebase they are supposedly git savvy enough to know what they are doing and then we really just need a "heads up" UX signalling that this is a commit message with some special meaning. So something like this seems to hit the right balance:
image

#KeepItSimple :)

**Comment by RussKie** [comment]
I like a different perspective, thanks @mdonatas.

Will a commit be treated as a fixup if the text changed though?

**Comment by gerhardol** [comment]
> Will a commit be treated as a fixup if the text changed though?

The subject after fixup must match. 
If you change it is something else, which is what the issue is about. 

As the button is implemented, I prefer the guidence it provides but I do not have a strong preference.

**Comment by mdonatas** [comment]
The subject must match. But that's the same if you are using `git` from the shell. If the user didn't know that subjects must match, this would be a good learning experience when the lines wouldn't get reordered in an interactive rebase. After all no harm is done at all.

_offtopic_
IMO there are way too many guard rails on by default already. For instance git embraces commit graph modifications with rebase and amending, there is even a saying `rebase makes it look as if you did everything perfect the first time` but GE scares you with dialogs about history rewrite which is one of the pillar features of git. And sadly very few go to change these settings.
I recommend GE to everyone because it exposes `git` features in GUI form without interpretations or restrictions as closely as possible meaning GE embraces `git` as opposed to crippling it. I am voting with both hands for continuing to embrace `git` as one of the GE pillar features. 🕊

**Comment by pmiossec** [comment]
>GE scares you with dialogs about history rewrite which is one of the pillar features of git.

The perfect thing would have been to display this message **only** when the commit has been already pushed.

At least we could update the warning message to explain that.

**Comment by RussKie** [comment]
> _offtopic_
> IMO there are way too many guard rails on by default already. For instance git embraces commit graph modifications with rebase and amending, there is even a saying `rebase makes it look as if you did everything perfect the first time` but GE scares you with dialogs about history rewrite which is one of the pillar features of git. And sadly very few go to change these settings.

Would you like to start a discussion about these?
I am getting tired of certain pop-ups and I would like to dial them down. Maybe even discuss creating "notice" vs "adept" modes which may have different levels of safeguards.

**Comment by RussKie** [comment]
> The perfect thing would have been to display this message **only** when the commit has been already pushed.

How do you see it implemented?

**Comment by RussKie** [comment]
> The subject must match.

If subject must match to the dot, it is not quite unreasonable to make the textbox readonly to indicate it to the user that it isn't something they should be changing.
However to allow the use-case where a user may change their mind and do another type of commit (i.e. not a fixup), we need to make the textbox editable.

If we leave the textbox editable for the special types of commits (i.e. fixup) our users may be left confused, change text and then wonder why they didn't get a fixup commit they wanted.

I'm speaking from first-hand experience where I had/have to teach engineers (including smart senior level devs and architects) about these git commands, and a lot of time I need(ed) to hold their hands with git. The same way others held my hand when I was learning git.

**Comment by mrexodia** [comment]
With regards to detecting if something is pushed, I guess you can check
where origin/branch is and only warn if the rebase is in the past of
origin/branch.

The way me (and my team) are using git is that we rebase and force push
feature branches until everything is clean so the message is useless anyway
(similar to the force push warning). For that the concept of “protected”
branches could be explored, but generally the server is enforcing it so
it’s probably not worth it.

**Comment by gerhardol** [comment]
>> offtopic

> Would you like to start a discussion about these?
I am getting tired of certain pop-ups and I would like to dial them down.

We should maybe open a separate issue to discuss this?
I do not mind a few OT posts but we should not go to far away from this change.
(As said before, I prefer guidance here, the old way to 'break-out' was not too bad for me.)

Traditional version control systems do not allow history rewriting so this is a change of mindset that does not come instantly. As Git is becoming the dominant VCS, I feel the trend is going to from banning rewrite via allowing private changes to enforcing rewrites. At work we require --ff-only, the history gets much more readable.

Maybe popup preset based on workflow is a way forward. Some warnings like amend should be reworded too.
Rules to follow the server settings like not rewrite protected branches like master should maybe be added too (that should be added to GitExtensions.settings for repo to be effective.)

**Comment by RussKie** [comment]
> We should maybe open a separate issue to discuss this?

That was my suggestion :) Sorry, I should have been a bit clearer.

File: GitUI/Translation/English.xlf [hunk 1]
```diff
@@ -3568,6 +3568,10 @@ You can unset the template:
         <source>Interactive Add</source>
         <target />
       </trans-unit>
+      <trans-unit id="modifyCommitMessageButton.Text">
+        <source>Modify this special commit m&amp;essage</source>
+        <target />
+      </trans-unit>
```
Comment (mstv) [inline GitUI/Translation/English.xlf]: The TranslationApp seems to ignore ToolTips. Must I really create a TranslationString by hand?

**Comment by RussKie** [comment]
Hmm... I vaguely remember we had issues with translations for tooltips, but
I don't remember whether it was resolved.
I speculate but I don't think we set tooltips in designer files.

**Comment by pmiossec** [comment]
I also don't remember but perhaps #3745 could help. Especially, https://github.com/gitextensions/gitextensions/issues/3745#issuecomment-306492000

**Comment by mstv** [comment]
Added tooltip translation and squashed.

File: GitUI/CommandsDialogs/FormCommit.Designer.cs [hunk 2]
```diff
@@ -1140,7 +1156,7 @@ private void InitializeComponent()
             this.CommitAndPush.Size = new System.Drawing.Size(171, 26);
             this.CommitAndPush.TabIndex = 9;
             this.CommitAndPush.TabStop = false;
-            this.CommitAndPush.Text = "C&ommit && push";
+            this.CommitAndPush.Text = "Commit && &push";
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.Designer.cs]: Changes like this look very sneaky. The PR's description briefly mentions this change, but it doesn't go into details as to what changed.
It won't be an easy/obvious bisect to find when and what changed, if users complain.

Please extract the changes to mnemonics in a separate commit and clearly describe the change.

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 3]
```diff
@@ -803,6 +789,21 @@ IReadOnlyList<GitItemStatus> GetAllChangedFilesWithSubmodulesStatus()
             }
         }
 
+        private void ConfigureMessageBox()
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs]:
```suggestion
        private void ConfigureMessageBox(CommitKind commitKind, bool useFormCommitMessage)
```

This way we can test it. Also please add tests.

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 4]
```diff
@@ -3297,25 +3330,6 @@ internal TestAccessor(FormCommit formCommit)
                 string commitMessageText, bool usingCommitTemplate, bool ensureCommitMessageSecondLineEmpty)
                 => FormCommit.FormatCommitMessageFromTextBox(commitMessageText, usingCommitTemplate, ensureCommitMessageSecondLineEmpty);
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs]:
```suggestion
                    => FormCommit.FormatCommitMessageFromTextBox(commitMessageText, usingCommitTemplate, ensureCommitMessageSecondLineEmpty);
```

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 5]
```diff
@@ -129,6 +129,9 @@ public sealed partial class FormCommit : GitModuleForm
         private readonly TranslationString _commitCommitterInfo = new TranslationString("Committer");
         private readonly TranslationString _commitCommitterToolTip = new TranslationString("Click to change committer information.");
 
+        private readonly TranslationString _modifyCommitMessageButtonToolTip
+            = new TranslationString("Enable the modification of this special commit message (Do it at your own risk!)");
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs]:
```suggestion
            = new TranslationString("If you change the commit message git will treat this commit as a ordinary commit (i.e. it may no longer be a fixup or an autosquash commit)");
```

I'm not sure if newline is rendered in tooltips, if it is - perhaps, add a newline before the left paren.

File: GitUI/CommandsDialogs/FormCommit.Designer.cs [hunk 6]
```diff
@@ -1033,12 +1034,27 @@ private void InitializeComponent()
...
+            //
+            this.modifyCommitMessageButton.AutoSize = true;
+            this.modifyCommitMessageButton.BackColor = System.Drawing.SystemColors.ButtonFace;
+            this.modifyCommitMessageButton.FlatStyle = System.Windows.Forms.FlatStyle.Flat;
+            this.modifyCommitMessageButton.ForeColor = System.Drawing.SystemColors.HotTrack;
+            this.modifyCommitMessageButton.Location = new System.Drawing.Point(185, 55);
+            this.modifyCommitMessageButton.Name = "modifyCommitMessageButton";
+            this.modifyCommitMessageButton.Size = new System.Drawing.Size(186, 25);
+            this.modifyCommitMessageButton.TabIndex = 9;
+            this.modifyCommitMessageButton.Text = "Modify this special commit m&essage";
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.Designer.cs]: It feels we're creating a new terminology here. I couldn't find any references to "special commit" (or "normal commits" for that matter).
Will it be less clear if we word as follows?
```suggestion
            this.modifyCommitMessageButton.Text = "Modify the commit m&essage";
```

File: GitUI/CommandsDialogs/FormCommit.cs (hunk 3 above)
Comment (mstv) [reply GitUI/CommandsDialogs/FormCommit.cs]: Pro: It emphasizes the depencies.
The drawback would be that all three callers must specify these options without being able to declare `CommitInfo` `static`. It could still evaluate other members.

I don't see any substantial benefit to add a test for a method like this.
Test cases would be only a guard against the modification of this very simple logic (in addition to the mandatory PR review).
```
       private void ConfigureMessageBox(CommitKind commitKind, bool useFormCommitMessage)
        {
            modifyCommitMessageButton.Visible = useFormCommitMessage && commitKind != CommitKind.Normal;
            bool messageCanBeChanged = useFormCommitMessage && commitKind == CommitKind.Normal;

            Message.Enabled = messageCanBeChanged;

            commitMessageToolStripMenuItem.Enabled = messageCanBeChanged;
            commitTemplatesToolStripMenuItem.Enabled = messageCanBeChanged;

            Message.WatermarkText = useFormCommitMessage
                ? _enterCommitMessageHint.Text
                : _commitMessageDisabled.Text;
        }
```

File: GitUI/CommandsDialogs/FormCommit.Designer.cs (hunk 6 above)
Comment (mstv) [reply GitUI/CommandsDialogs/FormCommit.Designer.cs]: It is a "special (commit) _message_". I understand your objection against introducing new terminology. The new tooltip describes it clear enough.

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 7]
```diff
@@ -3263,6 +3265,37 @@ private void Message_Enter(object sender, EventArgs e)
             }
         }
 
+        private void EnableEditSpecialCommitMessage()
+        {
+            _commitKind = CommitKind.Normal;
+            ConfigureMessageBox();
+        }
```
[the remark this answers is not in the document]
Comment (mstv) [reply GitUI/CommandsDialogs/FormCommit.cs]: Since `_useFormCommitMessage` is `readonly`, the set method of a new property `CommitKind` is the most correct replacement.

File: GitUI/CommandsDialogs/FormCommit.Designer.cs (hunk 2 above)
Comment (mstv) [reply GitUI/CommandsDialogs/FormCommit.Designer.cs]: Moved the submarines to dedicated PRs. In the future, I'll remember that mixing things - although related - makes it more difficult to review and to bisect.
Unfortunately, there's little support from Git developing things in parallel in the same working copy.

File: GitUI/CommandsDialogs/FormCommit.cs (hunk 3 above)
Comment (mstv) [reply GitUI/CommandsDialogs/FormCommit.cs]: I guess this has become obsolete with the new property `CommitKind`.
Your bad feeling regarding this method was justified, once more. 👍

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 8]
```diff
@@ -129,6 +129,10 @@ public sealed partial class FormCommit : GitModuleForm
         private readonly TranslationString _commitCommitterInfo = new TranslationString("Committer");
         private readonly TranslationString _commitCommitterToolTip = new TranslationString("Click to change committer information.");
 
+        private readonly TranslationString _modifyCommitMessageButtonToolTip
+            = new TranslationString("If you change the first line of the commit message, Git will treat this commit as an ordinary commit." + Environment.NewLine
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs]:
```suggestion
            = new TranslationString("If you change the first line of the commit message, git will treat this commit as an ordinary commit," + Environment.NewLine
```

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 9]
```diff
@@ -129,6 +129,10 @@ public sealed partial class FormCommit : GitModuleForm
...
+                                    + "I.e. it may no longer be a fixup or an autosquash commit.");
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs]:
```suggestion
                                    + "i.e. it may no longer be a fixup or an autosquash commit.");
```

**Comment by mstv** [comment]
Sorry for merging too early, I expected that the msftbot was configured to wait for at least one approval.
@RussKie, can the msftbot's default behavior be configured? Or must I always type something like
`merge 1 day after approval by @RussKie or after two approvals`

File: GitUI/CommandsDialogs/FormCommit.cs (hunk 9 above)
Comment (mdonatas) [reply GitUI/CommandsDialogs/FormCommit.cs]: How about this

> If commit message following squash! or fixup! will not match any commits in an interactive rebase this commit will not be marked for squashing or fixup.

The grammar or exact working can be changed but the idea is to not "invoke magic" by mentioning "regular" commits as if this is anything but. Fact is it's just a regular commit i.e. there is nothing in its metadata to say otherwise and only if you will proceed to do an interactive rebase the message _might_ have some effect.

**Comment by RussKie** [comment]
Yes, you need to write something like "@msftbot merge if @desired_user_name approves"
Have a look at [bot's guide](https://github.com/OfficeDev/office-ui-fabric-react/wiki/Advanced-auto-merge)

**Comment by mstv** [comment]
> Have a look at bot's guide

I had; and I had the hope the instantly handled "status: auto merge" could be tamed - by the repo admin at least.
When you used it, it wrote something like you had reset its configuration. That's why I thought the default behavior could be configured.
Perhaps it can be added to the PR template.
